pkg/controller/master/virtualmachine/vmi_network_controller.go

```diff
@@ -12,107 +13,179 @@ import (
 
 	ctlappsv1 "github.com/harvester/harvester/pkg/generated/controllers/apps/v1"
 	vmv1 "github.com/harvester/harvester/pkg/generated/controllers/kubevirt.io/v1"
+	util "github.com/harvester/harvester/pkg/util"
 )
 
 type VMNetworkController struct {
-	vmCache   vmv1.VirtualMachineCache
-	vmClient  vmv1.VirtualMachineClient
-	vmiClient vmv1.VirtualMachineInstanceClient
-	crCache   ctlappsv1.ControllerRevisionCache
-	crClient  ctlappsv1.ControllerRevisionClient
+	vmCache       vmv1.VirtualMachineCache
+	vmClient      vmv1.VirtualMachineClient
+	crCache       ctlappsv1.ControllerRevisionCache
+	crClient      ctlappsv1.ControllerRevisionClient
+	vmiController vmv1.VirtualMachineInstanceController
 }
 
-// SetDefaultNetworkMacAddress set the default mac address of networks using the initial allocated mac address from the VMI status,
-// since the most guest OS will use the initial allocated mac address of its DHCP config, and on Kubevirt the VM restart it will re-allocate
-// a new mac address which will lead the original network unreachable.
-func (h *VMNetworkController) SetDefaultNetworkMacAddress(id string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
-	if id == "" || vmi == nil || vmi.DeletionTimestamp != nil {
+func (h *VMNetworkController) SyncMacAddressAndControllerRevision(id string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	if id == "" || vmi == nil || vmi.DeletionTimestamp != nil || vmi.Status.Phase != kubevirtv1.Running || len(vmi.Status.Interfaces) == 0 {
 		return vmi, nil
 	}
 
-	if vmi.Status.Phase != kubevirtv1.Running {
-		return vmi, nil
+	// ensure vmi mac is backfilled to vm
+	retryvmi := false
+	if vmi, err := h.syncNetworkDynamicMacAddress(vmi, &retryvmi); err != nil {
+		return vmi, err
+	}
+
+	// ensure vmi mac is backfilled to rc
+	retrycr := false
+	if vmi, err := h.syncControllerRevision(vmi, &retrycr); err != nil {
+		return vmi, err
 	}
 
-	if len(vmi.Status.Interfaces) == 0 {
+	// vmi may end onChange, but vm/rc is not synced yet
```

Comment:
I am not sure I understand why vm and rc won't be sync'ed. Both `h.syncNetworkDynamicMacAddress()` and `h.syncControllerRevision` return with `Update()` completed. Or are we waiting for kubevirt side to sync up?

Reply:
yes, the generation is updated by apiserver when backfill MAC;  observedgeneration is updated by kubevirt controller, we need to wait kubevirt finishs its work

refer https://github.com/harvester/harvester/issues/6844#issuecomment-2551020204,